specs/core/0_beacon-chain.md

```diff
@@ -1264,50 +1263,56 @@ def get_genesis_beacon_state(genesis_validator_deposits: List[Deposit],
 
 ## Beacon chain state transition function
 
-We now define the state transition function. At a high level, the state transition is made up of four parts:
+The post-state corresponding to a pre-state `state` and a block `block` is defined as `state_transition(state, block)`. State transitions that trigger an unhandled exception (e.g. a failed `assert` or an out-of-range list access) are considered invalid.
 
-1. State caching, which happens at the start of every slot.
-2. The per-epoch transitions, which happens at the start of the first slot of every epoch.
-3. The per-slot transitions, which happens at every slot.
-4. The per-block transitions, which happens at every block.
```

Comment:
The `state_transition(state: BeaconState, block: BeaconBlock)` function assumes that to apply state transition, you need a `BeaconBlock`. But in the protocol, we should be able to apply per-slot transition without block. However, that's more difficult to present in code elegantly (with parameter realtime `slot`).

Reply:
Right. I think it's fine to define `state_transition` as a `(state: BeaconState, block: BeaconBlock) -> BeaconState` function. Two reasons:

* The per-slot state transition can be inferred by the per-block state transition. (This is similar to deserialisation being inferred from serialisation, or BLS signing being inferred from BLS verification.)
* We can clarify in the validator doc that validators are expected to run the per-slot state transition function in some instances.

Reply:
I think I lean toward

```python
def transition_to(state, slot):
    ...

def state_transition(state, block):
    transition_to(state, block.slot)
    process_block(state, block)
```

It is both readable and immediately useful in other areas of the spec. Right now, we kind of hand-wave how to transition through empty slots in the validator doc and the fork choice, but having this explicit function to point to would be more explicit. _Plus_ if we make the fork choice executable, we'll need this `transition_to` functionality

Reply:
Here's the new organisation for `state_transition`:

* `process_slots`
  * `process_slot`
  * `process_epoch`
    * `process_justification_and_finalization`
    * `process_crosslinks`
    * `process_rewards_and_penalties`
    * `process_registry_updates`
    * `process_slashings`
    * `process_final_updates`
* `process_block`
  * `process_block_header`
  * `process_randao`
  * `process_eth1_data`
  * `process_operations`
    * `process_proposer_slashing`
    * `process_attester_slashing`
    * `process_attestation`
    * `process_deposit`
    * `process_voluntary_exit`
    * `process_transfer`
  * `process_state_root`

Reply:
fwiw, we have two separate functions for block and blockless transtions in nimbus: https://github.com/status-im/nim-beacon-chain/blob/17fa4997556d0a69814144744e389cfa97c8f5ee/beacon_chain/state_transition.nim#L1233

part of the reason is that validation of block data and application of it is mixed up, so we will have modified the old state before knowing that the block is valid, which means we must also be able to roll back. blockless transitions however can't fail, so there's a difference / simplification in error handling as well.